rewrite-java/src/main/java/org/openrewrite/java/search/SemanticallyEqual.java

```diff
@@ -0,0 +1,184 @@
+
+/*
+ * Semantically Equal recursively checks the equality of each element of two ASTs to determine if the two trees are semantically equal.
+ * This is necessary because ASTs are so frequently recreated that merely comparing the IDs of two ASTs is ineffective.
+ * SemanticallyEqual has only been implemented for annotations for now.
+ */
+@Incubating(since = "6.0.0")
+public class SemanticallyEqual {
+
+    protected SemanticallyEqualProcessor processor;
+
+    public SemanticallyEqual() {
+        this.processor = new SemanticallyEqualProcessor();
+    }
+
+    public boolean areEqual(J firstElem, J secondElem) {
+        this.processor.visit(firstElem, secondElem); // returns null, but changes value of class variable isEqual
+        return this.processor.isEqual;
+    }
+
+    /**
+     * Note: The following visit methods extend JavaProcessor in order to inherit access to the
+     * visitor pattern set up there; however, the necessity to return a J did not fit the purposes of
+     * SemanticallyEqualProcessor, so while the equality is tracked in isEqual, all the visitors return null.
+     */
+    private static class SemanticallyEqualProcessor extends JavaProcessor<J> {
+        boolean isEqual;
+
+        SemanticallyEqualProcessor() {
+            isEqual = true;
+        }
+
+        @Override
+        public J visitAnnotation(J.Annotation firstAnnotation, J second) {
+            if(!(second instanceof J.Annotation)) {
+                isEqual = false;
+                return null;
+            }
+            J.Annotation secondAnnotation = (J.Annotation) second;
+
+            if(firstAnnotation.getArgs() != null && secondAnnotation.getArgs() != null) {
+                if(firstAnnotation.getArgs().getElem() != null &&
+                        secondAnnotation.getArgs().getElem() != null &&
+                        firstAnnotation.getArgs().getElem().size() == secondAnnotation.getArgs().getElem().size()) {
+
+                    List<JRightPadded<Expression>> firstArgs = firstAnnotation.getArgs().getElem();
+                    List<JRightPadded<Expression>> secondArgs = secondAnnotation.getArgs().getElem();
+
+                    for(int i = 0; i < firstArgs.size(); i++) {
+                        this.visit(firstArgs.get(i).getElem(), secondArgs.get(i).getElem());
+                    }
+                }
+                else {
+                    isEqual = false;
+                    return null;
+                }
+
+                this.visitTypeName(firstAnnotation.getAnnotationType(), secondAnnotation.getAnnotationType());
+            }
+            return null;
+        }
+
+        @Override
+        public J visitIdentifier(J.Ident firstIdent, J second) {
+            if(!(second instanceof J.Ident)) {
+                isEqual = false;
+                return null;
+            }
+            J.Ident secondIdent = (J.Ident) second;
+
+            isEqual &= Objects.equals(firstIdent.getType(), secondIdent.getType()) &&
```

Comment:
`&=` does not mean quite what you think it means.
`a &= b` is short for `a = a & b`.  `&` is bitwise-AND. `&&` is logical AND. It works out okay in this case because `boolean` is a single bit so bitwise has the same effect as logical and.

But you should still switch this and all similar statements to:
```java
isEqual = isEqual && ....
```
because logical AND short-circuits. If `isEqual` is `false` then the right side of the && wont even be evaluated, skipping a calculation whose conclusion is foregone.